1-Draft/RFCNNNN-Test-Connection.md

```diff
@@ -0,0 +1,121 @@
+---
+RFC: XXXX
+Author: Joel Sallow
+Status: Draft
+Version: 1.0
+Area: Cmdlets
+Comments Due: 6/10/19
+Plan to Implement: Yes
+---
+
+# Proposed Changes to `Test-Connection` Cmdlet
+
+## Syntax
+
+```
+# All Sets
+Test-Connection [-TargetName] <string[]> [-IPv4] [-IPv6] [-ResolveDestination] [-TimeoutSeconds <int>] [-Quiet] [<CommonParameters>]
+
+# Set 1
+[-Ping] [-Count <int>] [-Delay <int>] [-Source <string>] [-MaxHops <int>] [-BufferSize <int>] [-DontFragment]
+
+# Set 2
+[-Ping] [-Continues] [-Delay <int>] [-Source <string>] [-MaxHops <int>]  [-BufferSize <int>] [-DontFragment] 
+
+# Set 3
+Test-Connection [-TargetName] <string[]> -Traceroute [-Source <string>] [-MaxHops <int>] 
+
+# Set 4
+Test-Connection [-TargetName] <string[]> -MTUSizeDetect
+
+# Set 5
+Test-Connection [-TargetName] <string[]> -TCPPort <int> [-Source <string>]
+```
+
+## Changes
+
+1. All current host/information output is relegated to the `-Verbose` stream.
+2. No progress bar per committee decision in [PowerShell#6768](https://github.com/PowerShell/PowerShell/issues/6768).
+3. All output occurs as soon as data is available, individual records per ping reply / trace hop.
+4. All presented data should be readily available as properties to the user.
+
+## Output Type Proposed Members
+
+### PingStatus
+
+```csharp
+class PingStatus
+{
+    string Source { get; }
+    IPAddress Address { get; }
+    string Destination { get; }
+    int RoundtripTime { get; }
+    int BufferSize { get; }
+    PingOptions Options { get; }
+    // Hidden in the formatter when not set; only populated & shown when using the -MtuSizeDetect switch
+    int MtuSize { get; } = -1
+}
+```
+
+### TraceStatus
+
+```csharp
+class TraceStatus
+{
+    int Hop { get; }
+    int[] RoundTripTimes { get; }
+    string Destination { get; }
+    IPAddress DestinationAddress { get; }
+    IPAddress HopAddress { get; }
+    int BufferSize { get; }
+    PingStatus[] Replies { get; }
+}
+```
+
+## Mockups
+
+### `Test-Connection www.google.com`
+
+- Information from the `Replies` property of the output object should be included in the main output object, and primary output mode should be _multiple_ objects, each representing a single ping attempt/reply object.
+- Buffer _data_ is generally irrelevant, as it cannot be specified, and only a `BufferSize` property should be exposed. `Replies.Buffer` property should remain `private`.
+- `Replies.Options` property should be hidden from default formatting.
+- Results output as a table, grouped by Destination address (in the case that multiple destinations are specified).
+
+**Resulting output:**
+```code
+   Destination: www.google.com
+Source  Address       RoundtripTime (ms) BufferSize
```

Comment:
You can format the header so that it wraps:

```code
Roundtrip
Time (ms)
```

Personally, I'd rather use an extra row than have so much whitespace between columns

Reply:
Right, but that then implies a space in the property name, and we're unfortunately getting even farther from the actual structure of the object. 😕 

I learned PowerShell from Reddit at first; you would be surprised how many newer folks get tripped up on something seemingly so simple. 🙁

Reply:
That doesn't imply any whitespace as the formatter is splitting it across two rows.  I understand it might be confusing to new users, but I believe we already do this in some formatters.

Reply:
You're probably not wrong, I'm just saying I don't think it's an especially great pattern to continue. :confused:

Reply:
I'm with @vexx32 on this one, but I think a compromise would be just newline the units (which has the added benefit of making the units appear distinct from the property name).

Reply:
Sounds good! I also changed the property name to `Latency` as that... is just a lot easier to work with for formatting and I think is a bit clearer too.

Reply:
@SteveL-MSFT @joeyaiello do you guys have an example of how to setup a format header so it'll wrap? Currently I'm seeing things like this when it try:

![image](https://user-images.githubusercontent.com/32407840/59897361-51f93f00-93ba-11e9-91c9-16f193c3d101.png)

Note the extra space to the left of those columns -- the formatter is counting the space as though it's all still on one line, presumably _including_ the newline character.

If I don't insert an explicit newline and try to force it to wrap the header by restricting the column width, it simply refuses to enforce the width setting. If I try to set a width _and_ use a newline in the header

Reply:
Consider this resolved; I misunderstood the limitations of the `width` setting; I didn't realize that all items to the left needed the width specified.